feat: add v14 modals guide (#1060)
discordjs/guide: guide/interactions/modals.md

# Modals

> With modal interactions you can create pop-out forms that allow users to provide you with formatted input. We'll cover how to create, send and receive modals using discord.js!

Reviewer: ```suggestion
With modals, you can create pop-up forms that allow users to provide you, through submissions, with formatted inputs. We'll cover how to create, show, and receive modal forms using discord.js!
```

::: tip
This page is a follow-up to the [interactions (slash commands) page](/interactions/slash-commands.md). Please carefully read those first so that you can understand the methods used in this section.

> :::

Reviewer 2: This page is still in version 13 (not the problem here), so shouldn't the check `if (!interaction.isCommand()) return;` be `if (!interaction.isChatInputCommand()) return;`? The ping command is a chat input command.

> ## Building and Responding with Modals

Reviewer 3: ```suggestion
## Building and responding with modals
```

Unlike message components, modals aren't strictly components themselves. They're a callback structure used to respond to interactions.

::: tip

> You can have a maximum of five `ActionRow`s per modal, and one `TextInputComponent` within an `ActionRow`. Currently, you cannot use `SelectMenuComponent`s or `ButtonComponents` in modal action rows.

Reviewer 2: ```suggestion
You can have a maximum of five `ActionRow`s per modal, and one `TextInputComponent` within an `ActionRow`. Currently, you cannot use `SelectMenuComponent`s or `ButtonComponent`s in modal action rows.
```

> You can have a maximum of five `ActionRow`s per modal, and one `TextInputComponent` within an `ActionRow`. Currently, you cannot use `SelectMenuComponent`s or `ButtonComponent`s in modal action rows.

Reviewer 4: i am unsure if we want to change these

:::

> To create a modal you construct a new `Modal`. You can then use the setters to add the title.

Reviewer 4: ```suggestion
To create a modal you construct a new `ModalBuilder`. You can then use the setters to add the title.
```

```js {1,7-13}

> const { Modal } = require('discord.js');

Reviewer 4: ```suggestion
const { ModalBuilder } = require('discord.js');
```

client.on('interactionCreate', async interaction => {
	if (!interaction.isCommand()) return;

	if (interaction.commandName === 'ping') {

> 		const modal = new Modal()

Reviewer 4: ```suggestion
		const modal = new ModalBuilder()
```

			.setCustomId('myModal')
			.setTitle('My Modal');

> 		// TODO: Add components to modal...

Reviewer 5: ```suggestion
		// Add components to modal...
```

		await interaction.showModal(modal);
	}
});
```
::: tip

> The custom ID is a developer-defined string of up to 100 characters.

Reviewer 2: I'm pretty sure we should be lowercasing it. I've noticed some other places in the guide where "ID" is used, but that should probably be changed too?

```suggestion
The custom id is a developer-defined string of up to 100 characters.
```

:::

> As you can see we construct the modal and assign it a custom id and a title. Once our modal is constructed we send as a response to the interaction via `#showModal`. 

Reviewer 2: ```suggestion
As you can see we, construct the modal and assign it a custom id and a title. Once our modal is constructed, we send it as a response to the interaction via `#showModal`. 
```

We're still missing one step - adding inputs. Adding inputs is similar to adding components to messages.

```js {1,12-34}

> const { ActionRow, Modal, TextInputComponent, TextInputStyle } = require('discord.js');

Reviewer 4: ```suggestion
const { ActionRowBuilder, ModalBuilder, TextInputBuilder, TextInputStyle } = require('discord.js');
```

client.on('interactionCreate', async interaction => {
	if (!interaction.isCommand()) return;

	if (interaction.commandName === 'ping') {
		// Create our modal

> 		const modal = new Modal()

Reviewer 4: ```suggestion
		const modal = new ModalBuilder()
```

			.setCustomId('myModal')
			.setTitle('My Modal');

		// Add components to modal

		// Let's create our text inputs

> 		const favoriteColorInput = new TextInputComponent()

Reviewer 4: ```suggestion
		const favoriteColorInput = new TextInputBuilder()
```

			.setCustomId('favoriteColorInput')
		    // The label is the prompt the user sees for this input
			.setLabel("What's your favorite color?")
		    // Short means only a single line of text
			.setStyle(TextInputStyle.Short);

> 		const hobbiesInput = new TextInputComponent()

Reviewer 4: ```suggestion
		const hobbiesInput = new TextInputBuilder()
```

			.setCustomId('hobbiesInput')
			.setLabel("What's your favorites hobbies?")
		    // Paragraph means multiple lines of text.
			.setStyle(TextInputStyle.Paragraph);

		// An action row only holds one text input,
		// so we need one action row per text input.

> 		const firstActionRow = new ActionRow().addComponents(favoriteColorInput);

Reviewer 4: ```suggestion
		const firstActionRow = new ActionRowBuilder().addComponents(favoriteColorInput);
```

> 		const secondActionRow = new ActionRow().addComponents(hobbiesInput);

Reviewer 4: ```suggestion
		const secondActionRow = new ActionRowBuilder().addComponents(hobbiesInput);
```

		// Now we need to add our inputs into the modal
		modal.addComponents(firstActionRow, secondActionRow);

		// Show our modal

> 		await interaction.showModal(modal);

Reviewer 6: interaction.showModal is not a function

Author: It most definitely is, you most likely have a mistake somewhere in your code.

	}
});
```

Restart your bot and invoke the `/ping` command again. You should see a popup form resembling the image below:

TODO: Add modal image

> ## Receiving Modal Submissions

Reviewer 3: ```suggestion
## Receiving modal submissions
```

Modals are received via an interaction. You can check if a given interaction is a modal by invoking the `#isModalSubmit()` method:

```js {2}
client.on('interactionCreate', interaction => {
	if (!interaction.isModalSubmit()) return;
	console.log(interaction);
});
```

> ## Extracting Data from Modal Submissions

Reviewer 3: ```suggestion
## Extracting data from modal submissions
```

# Modals

With modals, you can create pop-up forms that allow users to provide you, with formatted inputs, through submissions. We'll cover how to create, show, and receive modal forms using discord.js!

::: tip

> This page is a follow-up to the [interactions (slash commands) page](/interactions/slash-commands.md). Please carefully read those first so that you can understand the methods used in this section.

Reviewer 7: ```suggestion
This page is a follow-up to the [interactions (slash commands) page](/interactions/slash-commands.md). Please carefully read that section first, so that you can understand the methods used in this section.
```

:::

## Building and responding with modals

Unlike message components, modals aren't strictly components themselves. They're a callback structure used to respond to interactions.

::: tip
You can have a maximum of five `ActionRowBuilder`s per modal builder, and one `TextInputBuilder` within an `ActionRowBuilder`. Currently, you cannot use `SelectMenuBuilder`s or `ButtonBuilder`s in modal action rows builders.
:::

> To create a modal you construct a new `ModalBuilder`. You can then use the setters to add the title.

Reviewer 7: ```suggestion
To create a modal you construct a new `ModalBuilder`. You can then use the setters to add the custom id and title.
```

```js {1,7-13}
const { ModalBuilder } = require('discord.js');

client.on('interactionCreate', async interaction => {
	if (!interaction.isChatInputCommand()) return;

	if (interaction.commandName === 'ping') {
		const modal = new ModalBuilder()
			.setCustomId('myModal')
			.setTitle('My Modal');

		// TODO: Add components to modal...

		await interaction.showModal(modal);
	}
});
```
::: tip
The custom id is a developer-defined string of up to 100 characters.
:::

> As you can see, we construct the modal and assign it a custom id and a title. Once our modal is constructed, we send it as a response to the interaction via `#showModal`.

Reviewer 7: ```suggestion
As you can see, you construct the modal and assign it a custom id and a title. After you added some user input elements, you will send it as a response to the interaction via `#showModal`.
```

be more explicit as to what's about to happen

> We're still missing one step - adding inputs. Adding inputs is similar to adding components to messages.

Reviewer 7: ```suggestion
You are still missing one of these steps - adding inputs. Adding inputs is similar to adding components to messages.
```

::: warning

> If you're using typescript you'll need to specify the type of components your action row holds. This can be done by specifying the
> generic parameter in `ActionRowBuilder`.

Reviewer 7: ```suggestion
If you're using typescript you'll need to specify the type of components your action row holds. This can be done by specifying the generic parameter in `ActionRowBuilder`.
```
obsolete line break

});
```
::: tip
The custom id is a developer-defined string of up to 100 characters.
:::

> As you can see we, construct the modal and assign it a custom id and a title. Once our modal is constructed, we send it as a response to the interaction via `#showModal`. 

Reviewer: ```suggestion
As you can see, we construct the modal and assign it a custom id and a title. Once our modal is constructed, we send it as a response to the interaction via `#showModal`.
```
